### ISSUE TITLE
Stripe Payment Element should not ask for country and postcode when the cart already knows them

### ISSUE THREAD

**Issue by thisisjamessmith** [craftcms/commerce-stripe#257]
### Description

When using `getPaymentFormHtml()` with the new Stripe Payment Element UI, there seems to be no practical way to instruct Stripe to hide and prepopulate its country and postcode fields if Craft's cart already has those details.

### Steps to reproduce

1. Have a Commerce cart that collects `billingAddress.countryCode` and `billingAddress.postalCode`. Choose UK as billing country.
2. Output the Stripe Payment Element UI via `gateway.getPaymentFormHtml(stripeParams)|raw`
3. Notice that the UI includes some blank fields for country and postcode.

As a workaround, we can at least prepopulate those values by passing the following parameters to `getPaymentFormHtml`:

```
{% set stripeParams = {
	...
	...
    elementOptions: {
        defaultValues: {
            billingDetails: {
                name: cart.customer.fullName ?? null,
                email: cart.email ?? null,
                phone: cart.billingAddress ? cart.billingAddress.addressPhone ?? null,
                address: {
                    line1: cart.billingAddress ? cart.billingAddress.addressLine1 ?? null,
                    line2: cart.billingAddress ? cart.billingAddress.addressLine2 ?? null,
                    city: cart.billingAddress ? cart.billingAddress.locality ?? null,
                    state: cart.billingAddress ? cart.billingAddress.locality ?? null,
                    postal_code: cart.billingAddress ? cart.billingAddress.postalCode ?? null,
                    country: cart.billingAddress ? cart.billingAddress.countryCode ?? null,
                }
            }
        }
    }
} %}

{% set gateway = craft.commerce.gateways.getGatewayById(1) %}
{{ gateway.getPaymentFormHtml(stripeParams)|raw }}
```

At a minimum, I would definitely expect Commerce to do that for us automatically, but having full control over Stripe's config would be even better.

This is a problem not only because the customer is confused why we're asking for some of their billing details again, but more importantly, it creates a potential inconsistency between the data we have collected and stored with that which the customer may have submitted in the final payment step. The billing info collected earlier by Commerce may be used in different circumstances, and creates a paper trail and accurate receipts and email notifications for both customer and store admins.

I understand that it's possible to "go manual" by not using `getPaymentFormHtml()` and writing all of our own code from scratch to instantiate the Payment Element, but I'm sure you'd agree that that's quite a significant undertaking, and Commerce already does a good job of getting us 90% of the way there - it's just not quite good enough yet.

### Additional info

- Craft CMS version: 4.5.6.1
- Craft Commerce version: 4.3.0
- Stripe for Craft Commerce version: 4.0.0
- PHP version: 8.1
- Database driver & version: MySQL 8

**Comment by johanneshakonsen** [comment]
Any chance you could post the entirety of your stripeParams? I can't get any response from stripe for the life of me except "Invalid payment or order". Almost like it's trying to do a payment on the initial connection to stripe

**Comment by thisisjamessmith** [comment]
I don't think that would really be relevant to this issue - better to ask in Discord, or open a new issue I think.

**Comment by joelzerner** [comment]
I've managed to hide the country and postcode fields by using the following code:

`{% set params = {
                  elementOptions: {
                    fields: {
                      billingDetails: {
                        address: {
                          country: 'never',
                          postalCode: 'never'
                        }
                      }
                    }
                  }
                } %}`

However I then get the following error when trying to submit a payment:

![image](https://github.com/craftcms/commerce-stripe/assets/11308275/ebbf618c-d847-4b7a-9e1d-fde6881b5fb8)
_paymentForm.js?v=1698326315:266 Uncaught (in promise) IntegrationError: You specified "never" for fields.billing_details.address.country when creating the payment Element, but did not pass confirmParams.payment_method_data.billing_details.address.country when calling stripe.confirmPayment(). If you opt out of collecting data via the payment Element using the fields option, the data must be passed in when calling stripe.confirmPayment()._

@lukeholder Could you advise the best way to integrate the default data here? We usually like to pass as much of the customer data to Stripe as possible - name, email, address etc.

This documentation for Stripe outlines adding a form event listener but not sure how we can do that in this instance:
https://stripe.com/docs/js/elements_object/create_payment_element#payment_element_create-customized_fields-fields

Craft Commerce Stripe documentation has some events to listen to but again not sure if these will do the trick. I've tried playing around with them but can't get past the error I am seeing.
https://github.com/craftcms/commerce-stripe#creating-a-stripe-payment-form

**Comment by lukeholder** [comment]
I will look to allow address to be passed into the payment form. I'll update here once it's ready.

**Comment by philipboomy** [comment]
Yeah passing in the name is important as Stripe uses this in their antifraud when they compare name to email. Right now we are getting a warning/message on all payment that name is missing etc.

Any update on this?

**Comment by lukeholder** [comment]
I'll look at this today and let you all know what we come up with.

**Comment by lukeholder** [comment]
We have a potential fix for this in this PR https://github.com/craftcms/commerce-stripe/pull/274

Stay tuned for the merge and release. Will update here once released.

**Comment by eshopsoftware** [comment]
> We have a potential fix for this in this PR #274
> 
> Stay tuned for the merge and release. I will update here once released.

Well anticipated. Thank you!

**Comment by nfourtythree** [comment]
Hi All

We have pushed up a fix for this that has been included in the latest release of the plugin `4.1.0`.

The README has been updated with instructions on how to pass the order (cart) to the payment form to get all the new functionality working.

Thanks!

**Comment by clenia1e** [comment]
Hi @nfourtythree  @lukeholder  I'm using this version   "@stripe/stripe-js": "4.1.0",
 I tried to  hide the postal code and country from the credit card form and I'm getting the same error 
 
`
          options={{
            fields: {
              billingDetails: {
                address: {
                  country: 'never',
                  postalCode: 'never',
                },
              },
            },
          }}
`

error:

`You specified "never" for fields.billing_details.address.country when creating the payment Element, but did not pass confirmParams.payment_method_data.billing_details.address.country when calling stripe.confirmPayment(). If you opt out of collecting data via the payment Element using the fields option, the data must be passed in when calling stripe.confirmPayment().`

**Comment by zesda** [comment]
It appears that this has not addressed the underlying issue in that `stripe.confirmPayment()` expects `confirmParams.payment_method_data.billing_details.address` to be provided, when `"never"` is specified for `fields.billing_details.address` (or for it's children).

Specifying `defaultValues.billingDetails.address` in the `elementOptions` seems to make no difference (as seen in the attached screenshot).

Screenshot is with is with `"craftcms/commerce-stripe": "5.0.4.2"` using the default Craft Commerce example templates.

![Image](https://github.com/user-attachments/assets/906303f2-40c2-4451-bda5-4765ce5b2ebc)

**Comment by philipboomy** [comment]
I have also emailed support as I cannot get it working proper. Stripe is still missing name which makes the fraud detection less efficient. I have not heard back.

**Comment by lukeholder** [comment]
@zesda Can you please give me the params you pass to getPaymentForm, any events you are using to modify the stripe request. Just the steps to reproduce, and I will take a look again.

**Comment by zesda** [comment]
@lukeholder Details as follows:

The only amendment was to `vendor/craftcms/commerce/example-templates/dist/shop/checkout/payment.twig` with updated `params` for the `craft\commerce\stripe\gateways\PaymentIntents` gateway (note the addition of `fields`):

```
{% set params = {
  appearance: {
    theme: 'stripe'
  },
  elementOptions: {
    layout: {
      type: 'accordion',
      defaultCollapsed: false,
      radios: false,
      spacedAccordionItems: false
    },
    fields: {
      billingDetails: {
        address: 'never'
      }
    }
  },
  order: cart,
  submitButtonClasses: 'cursor-pointer rounded px-4 py-2 inline-block bg-blue-500 hover:bg-blue-600 text-white hover:text-white my-2',
  submitButtonText: 'Pay',
  errorMessageClasses: 'bg-red-200 text-red-600 my-2 p-2 rounded',
} %}
```

From there, it was just a case of following the example checkout funnel with any combination of user, cart, addresses etc and I would receive this error when clicking "Pay" on the payment form (in the "Payment" page).

Following the trace from the JS error it pointed to the `vendor/craftcms/commerce-stripe/src/web/assets/elementsform/js/paymentForm.js` file on Line 323 (for `"craftcms/commerce-stripe": "5.0.4.2"`) where `this.stripeInstance.confirmPayment()` is called, it appears there's no accommodation for passing `confirmParams.payment_method_data.billing_details.address.*` in this scenario.

**Comment by joelzerner** [comment]
We're experiencing the same issue as @zesda
Really keen to get that resolved so we can release this into the wild!

We're also experiencing another issue related to Line 242 in `vendor/craftcms/commerce-stripe/src/web/assets/elementsform/js/paymentForm.js`
`if(json.redirect && json.cart.isCompleted === true) {`

Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'isCompleted')

Looks like json.cart doesn't exist. I can see json.order which has an isCompleted variable in it. Switching json.cart to json.order fixes the issue. Thoughts?

**Comment by joelzerner** [comment]
@lukeholder Any updates on this one?

**Comment by philipboomy** [comment]
An update on this would be appreciated. I cannot get name to appear in Stripe and it is needed for better fraud detection.

**Comment by lukeholder** [comment]
@joelzerner 

> We're also experiencing another issue related to Line 242 in `vendor/craftcms/commerce-stripe/src/web/assets/elementsform/js/paymentForm.js` `if(json.redirect && json.cart.isCompleted === true) {`

The above was resolved in 5.0.4.4


As for the address data, setting it to 'never' will not work since the data would then need to be provided in the confirm as per the Stripe docs.

I am not sure what the best course of action would be, to allow the developer to add confirmParams to the payment form, which would be most flexible (which would allow users to edit the HTML/`data-`) or check to see if you pass 'never' for the element options and pull those attributes from the order automatically?

**Comment by lukeholder** [comment]
@philipboomy 

> An update on this would be appreciated. I cannot get name to appear in Stripe and it is needed for better fraud detection.

We auto populate the billing info so the name should be there. Can you take a screeshot/share your payment form config and where you expect to see the name show up?

**Comment by joelzerner** [comment]
@lukeholder 
Thanks for the fix.

Either option sounds good to me. If the flexible one is achievable to get up and running quickly, that would be great. But happy with the other one too.

**Comment by RitterKnightCreative** [comment]
> check to see if you pass 'never' for the element options and pull those attributes from the order automatically?

@lukeholder I think this would work the best as well.  

FWIW there might also be a third option.  

If Commerce is trying to abstract as much of the Stripe API away from the developer, ,have a Commerce-specific param that sets the Stripe data that Commerce already has on the order and automatically pass in "never" for the values it already knows about so the developer doesn't have to do that (eg ``hideRedundantPaymentFields`` or something.)

From a DX perspective, I was under the impression from reading the docs that if you pass in the cart, Commerce would have everything it already needs to make the proper Stripe element already. It's only after testing that when I hit "Pay" I'm hit with a JS error.
